### PR TITLE
Fix for Unloaded ammo belt doesn't disappear

### PR DESCRIPTION
**Pull request by Firestorm01X2**
Fix for Unloaded ammo belt doesn't disappear

#### Summary
Bugfixes https://github.com/cataclysmbnteam/Cataclysm-BN/issues/49

#### Purpose of change
Fixes https://github.com/cataclysmbnteam/Cataclysm-BN/issues/49

#### Describe the solution
Bug happens only if player Unloads ammo belt by using U command (not from inventory). But belt removal was used only for avatar action. Solution is move belt removal to player::unload.

Partially implementing this: https://github.com/CleverRaven/Cataclysm-DDA/pull/40541

Transofrm 

`bool player::unload( item &it)`
to
`bool player::unload( item_location loc )`

Do essesary modfications to all references

Add "MAG_DESTROY" item removal to magazine handling in
`bool player::unload( item_location loc )`

Remove "MAG_DESTROY" removal from
`avatar_action::unload( avatar &you )`

#### Testing
1) Spawn 5.56x45mm ammo belt
2) Unload it from item menu and form U menu
3) In both cases you must get only ammo and linkages without emepty belt.

### CODE REVIEW

**Comment by Firestorm01X2** [comment]
Solution is working. Wating for approval. Only astyling left.

File: src/avatar_action.cpp [hunk 1]
```diff
@@ -1284,9 +1285,11 @@ void avatar_action::unload( avatar &you )
...
-    if( you.unload( *loc ) ) {
-        if( loc->has_flag( "MAG_DESTROY" ) && loc->ammo_remaining() == 0 ) {
-            loc.remove_item();
-        }
-    }
+    //if( you.unload( *loc ) ) {
+    //    if( loc->has_flag( "MAG_DESTROY" ) && loc->ammo_remaining() == 0 ) {
+    //        loc.remove_item();
+    //    }
+    //}
```
Comment (Coolthulhu) [inline src/avatar_action.cpp:1292]: Don't leave commented code.

File: src/game.cpp [hunk 2]
```diff
@@ -8652,9 +8652,9 @@ void game::reload_weapon( bool try_everything )
     reload_item();
 }
 
-bool game::unload( item &it )
+bool game::unload( item_location &loc )
```
Comment (Coolthulhu) [inline src/game.cpp:8655]: `item_location`s are usually passed by value, not reference. This allows simplifying some calls.

File: src/player.cpp [hunk 3]
```diff
@@ -3095,8 +3095,18 @@ bool player::add_or_drop_with_msg( item &it, const bool unloading )
     return true;
 }
 
-bool player::unload( item &it )
+bool player::unload( item_location &loc )
 {
+    unload_internal( loc );
```
Comment (Coolthulhu) [inline src/player.cpp:3100]: Is there a reason you have this as a separate function?
It's only used here.

Comment (Firestorm01X2) [reply src/player.cpp:3100]: Only to clear "MAG_DESTROY" items in single place. Otherwise I have to refactor player::unload_internal (old player::unload)

Comment (Coolthulhu) [reply src/player.cpp:3100]: You could just inline it. Paste the contents of the whole `unload_internal` here and remove the extra function.
Every single call to `unload` starts with a call to `unload_internal`, but no other function uses `unload_internal`.

**Comment by Coolthulhu** [comment]
I don't mean "inline" as in the keyword, just cut the contents of the function and place it where you'd call the function.

**Comment by Firestorm01X2** [comment]
Ready.

File: src/game.cpp (hunk 2 above)
Comment (Coolthulhu) [reply src/game.cpp:8655]: This one should be by-value too.

File: src/player.cpp [hunk 4]
```diff
@@ -3095,183 +3095,192 @@ bool player::add_or_drop_with_msg( item &it, const bool unloading )
...
+            target->type->invoke( *this, *target, this->pos() );
+        }
 
-    // Turn off any active tools
-    if( target->is_tool() && target->active && target->ammo_remaining() == 0 ) {
-        target->type->invoke( *this, *target, this->pos() );
+        add_msg( _( "You unload your %s." ), target->tname() );
+        return true;
+    };
+    bool result = unload_internal( loc );
```
Comment (Coolthulhu) [inline src/player.cpp:3276]: Oh, I see why you had it all as a function. The return value.
It's still better like this, so it's OK.

The ideal version would be to find the place where ammo belts are handled and place the `MAG_DESTROY` check in there.
From what I can tell, there is only one such place in the function.
The split between unload belt and destroy belt only made sense when `unload` took an `item&` argument.

File: src/player.cpp [hunk 5]
```diff
@@ -3095,183 +3095,192 @@ bool player::add_or_drop_with_msg( item &it, const bool unloading )
...
+                }
+                add_msg( _( "You disassemble your %s." ), it.tname() );
             } else {
-                add_msg( m_info, _( "You can't remove partially depleted plutonium!" ) );
+                add_msg( _( "You unload your %s." ), it.tname() );
+            }
+
+            mod_moves( -std::min( 200, mv ) );
+
+            return true;
```
Comment (Coolthulhu) [inline src/player.cpp:3214]: The section above this line is the only place where the magazine itself is unloaded. So this is the only place where an item with `MAG_DESTROY` has a chance to be triggered.

File: src/player.cpp (hunk 4 above)
Comment (Firestorm01X2) [reply src/player.cpp:3276]: Belts handled here:

```
            // remove the belt linkage
            if( it.is_ammo_belt() ) {
                if( it.type->magazine->linkage ) {
                    item link( *it.type->magazine->linkage, calendar::turn, qty );
                    add_or_drop_with_msg( link, true );
                }
                add_msg( _( "You disassemble your %s." ), it.tname() );
            } else {
                add_msg( _( "You unload your %s." ), it.tname() );
            }
```

But I am not sure the belts are the only case

File: src/player.cpp (hunk 5 above)
Comment (Firestorm01X2) [reply src/player.cpp:3214]: Got it

**Comment by Firestorm01X2** [comment]
Ready.

**Pull request merged**